src/core/Akka.Tests/Actor/ActorSystemSpec.cs

```diff
@@ -236,5 +339,105 @@ public void ScheduleTellOnce(TimeSpan delay, ICanTell receiver, object message,
         public TimeSpan HighResMonotonicClock { get; private set; }
         public IAdvancedScheduler Advanced { get; private set; }
     }
+
+    public class Wave : ReceiveActor
+    {
+        private IActorRef _master = Nobody.Instance;
+        private readonly HashSet<IActorRef> _terminaters = new HashSet<IActorRef>();
+
+        public Wave()
+        {
+            Receive<int>(n =>
+            {
+                _master = Sender;
+
+                for (int i = 0; i < n; i++)
+                {
+                    var man = Context.Watch(Context.System.ActorOf(Props.Create<Terminater>()));
+                    man.Tell("run");
+                    _terminaters.Add(man);
+                }
+            });
+
+            Receive<Terminated>(t =>
+            {
+                var child = t.ActorRef;
+
+                if (_terminaters.Contains(child))
+                    _terminaters.Remove(child);
+
+                if (_terminaters.Count == 0)
+                {
+                    _master.Tell("done");
+                    Context.Stop(Self);
+                }
+            });
+        }
+
+        protected override void PreRestart(Exception reason, object message)
+        {
+            if (!_master.IsNobody())
+                _master.Tell($"failed with {reason} while processing {message}");
+
+            Context.Stop(Self);
+        }
+    }
+
+    public class Terminater : ReceiveActor
+    {
+        public Terminater()
+        {
+            Receive<string>(s => "run".Equals(s), s => Context.Stop(Self));
+        }
+    }
+
+    public class SlowDispatcher : MessageDispatcherConfigurator
```

Comment:
Please review this carefully, I really don't know if this is implemented correctly. 
You can find the corresponding scala implementation [here](https://github.com/akka/akka/blob/master/akka-actor-tests/src/test/scala/akka/actor/ActorSystemSpec.scala#L103)